lib/aws-record/record/batch_writer.rb

```diff
@@ -0,0 +1,75 @@
+module Aws
+  module Record
+    class BatchWriter < Batch
+      include Enumerable
+
+      # This method calls
+      # {http://docs.aws.amazon.com/sdkforruby/api/Aws/DynamoDB/Client.html#batch_write_item-instance_method Aws::DynamoDB::Client#batch_write_item}},
+      # on the items stored in this BatchWriter. Note that this method does
+      # not support delete_request, only put_request and that only items
+      # of the same class can be written using this method. The following
+      # failure scenarios are prevented before being sent to DynamoDB:
+      #
+      #   - One or more tables specified in the BatchWriteItem request does not
+      #      exist.
+      #   - You try to perform multiple operations on the same item in the same
+      #      BatchWriteItem request. For example, you cannot put and delete the
+      #      same item in the same BatchWriteItem request.
+      #   - There are more than 25 requests in the batch.
+      #   - Primary key attributes specified on an item in the request do not
+      #      match those in the corresponding table's primary key schema.
+      #
+      # This method will not prevent the following failure scenarios and will
+      # need to be handled separately:
+      #
+      #   - Any individual item in a batch exceeds 400 KB.
+      #   - The total request size exceeds 16 MB.
+      #
+      # @param [Hash] opts options to pass on to the client call to
+      #   +#batch_item_write+. See the documentation above in the AWS SDK for
+      #   Ruby V2.
+      # @opts retry_count the number of retries to attempt on failed write
+      # @return
+      def save!(opts = {})
+        validate_items
+        @items.each_slice(25) do
+          batch_write_items(opts)
+        end
+      end
+
+      private
+      def batch_write_items(opts)
+        try_count = opts.delete(:retry_count) { |_| 4 }
```

Comment:
When you do look at non-retried errors from the client to handle, there are two cases that come to mind:

* `Aws::DynamoDB::Errors::ProvisionedThroughputExceededException`: In the case where the entire batch fails to write.
* Successful response, but `:unprocessed_items` is not empty: You'll want to roll those items over into your next batched write. Otherwise, you'll silently fail to write a subset of items, and users will have no visibility into what failed to write.

I'm not sure off the top of my head how aggressive retries should be for this, but it shouldn't be infinite, which means this collection should have a way to surface what parts of the enumeration failed to be written.

Further discussion: https://github.com/aws/aws-sdk-ruby/issues/1154